libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,319 @@
+func bindHostPorts(cfg []*types.PortBinding, proxyPath string) ([]*portBinding, error) {
+	if len(cfg) == 0 {
+		return nil, nil
 	}
 
 	// Try up to maxAllocatePortAttempts times to get a port that's not already allocated.
+	var err error
 	for i := 0; i < maxAllocatePortAttempts; i++ {
-		if host, err = portmapper.MapRange(container, bnd.HostIP, int(bnd.HostPort), int(bnd.HostPortEnd), ulPxyEnabled); err == nil {
-			break
+		var b []*portBinding
+		b, err = attemptBindHostPorts(cfg, cfg[0].Proto.String(), cfg[0].HostPort, cfg[0].HostPortEnd, proxyPath)
+		if err == nil {
+			return b, nil
 		}
 		// There is no point in immediately retrying to map an explicitly chosen port.
-		if bnd.HostPort != 0 && bnd.HostPort == bnd.HostPortEnd {
-			log.G(context.TODO()).Warnf("Failed to allocate and map port %d-%d: %s", bnd.HostPort, bnd.HostPortEnd, err)
+		if cfg[0].HostPort != 0 && cfg[0].HostPort == cfg[0].HostPortEnd {
+			log.G(context.TODO()).Warnf("Failed to allocate and map port: %s", err)
 			break
 		}
 		log.G(context.TODO()).Warnf("Failed to allocate and map port: %s, retry: %d", err, i+1)
 	}
+	return nil, err
+}
+
+// Allow unit tests to supply a dummy StartProxy.
+var startProxy = portmapper.StartProxy
+
+// attemptBindHostPorts allocates host ports for each port mapping that requires
+// one, and reserves those ports by starting docker-proxy.
+//
+// If the allocator doesn't have an available port in the required range, or the
+// docker-proxy process doesn't start (perhaps because another process has
+// already bound the port), all resources are released and an error is returned.
+// When ports are successfully reserved, a portBinding is returned for each
+// mapping.
+func attemptBindHostPorts(
+	cfg []*types.PortBinding,
+	proto string,
+	hostPortStart, hostPortEnd uint16,
+	proxyPath string,
+) (_ []*portBinding, retErr error) {
+	addrs := make([]net.IP, 0, len(cfg))
+	for _, c := range cfg {
+		addrs = append(addrs, c.HostIP)
+	}
+
+	pa := portallocator.Get()
+	port, err := pa.RequestPortsInRange(addrs, proto, int(hostPortStart), int(hostPortEnd))
 	if err != nil {
-		return err
+		return nil, err
 	}
+	defer func() {
+		if retErr != nil {
+			for _, a := range addrs {
+				pa.ReleasePort(a, proto, port)
+			}
+		}
+	}()
 
-	// Save the host port (regardless it was or not specified in the binding)
-	switch netAddr := host.(type) {
-	case *net.TCPAddr:
-		bnd.HostPort = uint16(host.(*net.TCPAddr).Port)
-		return nil
-	case *net.UDPAddr:
-		bnd.HostPort = uint16(host.(*net.UDPAddr).Port)
-		return nil
-	case *sctp.SCTPAddr:
-		bnd.HostPort = uint16(host.(*sctp.SCTPAddr).Port)
-		return nil
-	default:
-		// For completeness
-		return ErrUnsupportedAddressType(fmt.Sprintf("%T", netAddr))
+	res := make([]*portBinding, 0, len(cfg))
+	for _, c := range cfg {
+		cc := c.GetCopy()
+		pb := &portBinding{PortBinding: &cc}
+		pb.proxy, err = startProxy(c.Proto.String(), c.HostIP, port, c.IP, int(c.Port), proxyPath)
```

Comment:
Ah, yes - the interface is always 'up' at this point. And for network-connect or live-restore, the container task will already be running. It's probably worth re-jigging, although I think that may leave scope for bugs that vanish iptables rules - I'll take a look.

Reply:
I think it's ok as it is ... anything that happens before the iptables rules are added will hit the `FORWARD DROP` action in the filter rule. When the rules are added, the container's port is opened.

For live-restore / network connect, iptables rules are already in place when the proxy is started.

Reply:
Discussed with @akerouanton ... the problem is `docker-proxy` will accept a connection that would have passed it by if the iptables rules were in-place. So, although it won't be able to create the container-side connection until the iptables forwarding rule exists, the connection will be abandoned.

It's a small window, and a trade-off between creating rules for a port that belongs to something else (disrupting traffic on an already in-use host port, and possibly breaking it completely when rules are removed).

Perhaps we can modify `docker-proxy` so that it doesn't actually accept connections until it's told the iptables rules are set up ... needs investigation.